### PR TITLE
[Tools]: MAE Reconstructed Image Visualization

### PR DESCRIPTION
**Pull request by soonera**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Support MAE reconstructed image  visualization

## Modification

(1)  Add init_model() function to Initialize a model from checkpoint.
(2)  Add inference_model() function to infer an image with the loaded model.
(3)  Add forward_test() in Class MAE.
(4)  Add unpatchify() in Class MAEPretrainHead.
(5) Add mae_visualization.py and mae_visualization.ipynb

## BC-breaking (Optional)

Does the modification introduce changes that break the backward compatibility of the downstream repositories?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here and update the documentation.

## Checklist

**Before PR**:

- [ ] Pre-commit or other linting tools are used to fix the potential lint issues.
- [ ] Bug fixes are fully covered by unit tests, the case that causes the bug should be added in the unit tests.
- [ ] The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
- [ ] The documentation has been modified accordingly, like docstring or example tutorials.

**After PR**:

- [ ] If the modification has potential influence on downstream or other related projects, this PR should be tested with those projects, like MMDet or MMSeg.
- [ ] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

File: mmselfsup/apis/inference.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+    model.eval()
+    return model
+
+
+def inference_model(model, data):
+    """Inference a image with the model.
+
+    Args:
+        model (nn.Module): The loaded model.
+        data (PIL): The loaded image.
```
Comment (fangyixiao18) [inline mmselfsup/apis/inference.py:49]: PIL -> PIL.Image

File: mmselfsup/apis/inference.py [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+    Returns:
+        data (Tensor): The loaded image to input model.
```
Comment (fangyixiao18) [inline mmselfsup/apis/inference.py:51]:
```suggestion
        data (torch.Tensor): The loaded image to input model.
```

File: mmselfsup/models/heads/mae_head.py [hunk 3]
```diff
@@ -34,6 +37,20 @@ def patchify(self, imgs):
         x = x.reshape(shape=(imgs.shape[0], h * w, p**2 * 3))
         return x
 
+    def unpatchify(self, x):
+        """
+        x: (N, L, patch_size**2 *3)
+        imgs: (N, 3, H, W)
```
Comment (fangyixiao18) [inline mmselfsup/models/heads/mae_head.py:43]: please follow the format of the docstring
```suggestion
        Args:
            ...
        Returns:
            ...
```

File: mmselfsup/models/heads/mae_head.py [hunk 4]
```diff
@@ -24,7 +24,10 @@ def __init__(self, norm_pix=False, patch_size=16):
         self.patch_size = patch_size
 
     def patchify(self, imgs):
-
+        """
```
Comment (fangyixiao18) [inline mmselfsup/models/heads/mae_head.py:27]: check the format of docstring

**Review by fangyixiao18 — CHANGES_REQUESTED** [review]

**Comment by fangyixiao18** [comment]
Please also add tutorial for the using of MAE visualization

File: mmselfsup/apis/inference.py [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+
+def init_model(config, checkpoint=None, device='cuda:0', options=None):
+    """Initialize a model from config file.
+
+    Args:
+        config (str or :obj:`mmcv.Config`): Config file path or the config
+            object.
+        checkpoint (str, optional): Checkpoint path. If left as None, the model
+            will not load any weights.
+        options (dict): Options to override some settings in the used config.
```
Comment (YuanLiuuuuuu) [inline mmselfsup/apis/inference.py:21]:
```suggestion
        options (dict, optional): Options to override some settings in the used config.
```

File: mmselfsup/apis/inference.py [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+
+from mmselfsup.datasets import PIPELINES
+from mmselfsup.models import build_algorithm
+
+
+def init_model(config, checkpoint=None, device='cuda:0', options=None):
+    """Initialize a model from config file.
+
+    Args:
+        config (str or :obj:`mmcv.Config`): Config file path or the config
```
Comment (YuanLiuuuuuu) [inline mmselfsup/apis/inference.py:17]:
```suggestion
        config (str or :obj:``mmcv.Config``): Config file path or the config
```

File: mmselfsup/apis/inference.py [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+            object.
+        checkpoint (str, optional): Checkpoint path. If left as None, the model
```
Comment (YuanLiuuuuuu) [inline mmselfsup/apis/inference.py:19]: Missing docstring for ``device``

File: mmselfsup/apis/inference.py [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+            will not load any weights.
+        options (dict): Options to override some settings in the used config.
+    Returns:
+        nn.Module: The constructed model.
```
Comment (YuanLiuuuuuu) [inline mmselfsup/apis/inference.py:23]:
```suggestion
        nn.Module: The initialized model.
```

File: mmselfsup/apis/inference.py [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+    """
+    if isinstance(config, str):
+        config = mmcv.Config.fromfile(config)
+    elif not isinstance(config, mmcv.Config):
+        raise TypeError('config must be a filename or Config object, '
+                        f'but got {type(config)}')
+    if options is not None:
+        config.merge_from_dict(options)
+    # config.model.pretrained = None
```
Comment (YuanLiuuuuuu) [inline mmselfsup/apis/inference.py:32]: This line should be removed

File: mmselfsup/apis/inference.py [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+        # Mapping the weights to GPU may cause unexpected video memory leak
+        # which refers to https://github.com/open-mmlab/mmdetection/pull/6405
+        checkpoint = load_checkpoint(model, checkpoint, map_location='cpu')
+    model.cfg = config  # save the config in the model for convenience
+    model.to(device)
+    model.eval()
+    return model
+
+
+def inference_model(model, data):
```
Comment (YuanLiuuuuuu) [inline mmselfsup/apis/inference.py:44]: missing typehint

File: mmselfsup/apis/inference.py [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+    """Inference a image with the model.
```
Comment (YuanLiuuuuuu) [inline mmselfsup/apis/inference.py:45]:
```suggestion
    """Inference an image with the model.
```

**Review by YuanLiuuuuuu — COMMENTED** [review]
Please also add the return typehint. In addition, better provide a toy example (initial image, masked image and reconstructed image), showing the visualization results.
